### PR TITLE
update (passive_anticheat): Monthly alignment

### PR DESCRIPTION
**Pull request by acidmanifesto**
**Changes proposed**:

- Counter Measures, which breaks select cheat apps and client's connections when in use.
- Moved exsisting presistent countermeasures to bools. (Time Manipulation)
- Added additional countermeasures (teleport\flyhack)
- Logging via conf
- Counter Measures via conf
- Counter Measures Alerts via conf
- New trinitycore_string via world sql to be applied.
- Updated Character reports with counter measure columns.
- GM .anticheat player now shows counter measure count.
- Added new method MoveFallPlayer() to assist with counter measures

New Counter Measures Deployed reporting per player's current session
image

In game alerts for GM both screen and chat spam
image

Video

https://user-images.githubusercontent.com/16887899/228910878-dfd4f461-dd6d-4317-b263-0a9a45879ee4.mp4

New log examples of counter measures
```
2023-03-30_12:55:19 INFO  [anticheat] AnticheatMgr:: Speed-Hack (Speed Movement at 49 above allowed Server Set rate 7.) detected player Yarillee (GUID Full: 0x0000000000000004 Type: Player Low: 4) - Latency: 0 ms - Cheat Flagged at: .go xyz -8804.487305 686.484253 102.187088 0 3.712932
2023-03-30_12:55:19 INFO  [anticheat] AnticheatMgr:: Teleport-Hack detected player Yarillee (GUID Full: 0x0000000000000004 Type: Player Low: 4) - Latency: 0 ms - Cheat Flagged at: .go xyz -8598.000000 527.000000 109.000000 0 3.712932
2023-03-30_12:55:19 INFO  [anticheat] ANTICHEAT COUNTER MEASURE:: Yarillee TELEPORT HACK REVERTED PLAYER BACK TO .go xyz -8804.487305 686.484253 102.187050 0 3.712932
2023-03-30_12:55:30 INFO  [anticheat] AnticheatMgr:: Fly-Hack detected player Yarillee (GUID Full: 0x0000000000000004 Type: Player Low: 4) - Latency: 0 ms - Cheat Flagged at: .go xyz -8809.514648 683.804810 100.819611 0 3.606904
2023-03-30_12:55:30 INFO  [anticheat] ANTICHEAT COUNTER MEASURE:: Fly Hack detected player Yarillee (GUID Full: 0x0000000000000004 Type: Player Low: 4) - SMSG_MOVE_UNSET_CAN_FLY Set 
2023-03-30_12:55:48 INFO  [anticheat] AnticheatMgr:: Fly-Hack detected player Yarillee (GUID Full: 0x0000000000000004 Type: Player Low: 4) - Latency: 0 ms - Cheat Flagged at: .go xyz -8828.438477 690.385559 98.331505 0 2.004691
2023-03-30_12:55:48 INFO  [anticheat] ANTICHEAT COUNTER MEASURE:: Fly Hack detected player Yarillee (GUID Full: 0x0000000000000004 Type: Player Low: 4) - SMSG_MOVE_UNSET_CAN_FLY Set 
2023-03-30_12:56:54 INFO  [anticheat] AnticheatMgr:: Fly-Hack detected player Yarillee (GUID Full: 0x0000000000000004 Type: Player Low: 4) - Latency: 1 ms - Cheat Flagged at: .go xyz -8854.118164 742.426025 101.743782 0 0.241471
2023-03-30_12:56:54 INFO  [anticheat] ANTICHEAT COUNTER MEASURE:: Fly Hack detected player Yarillee (GUID Full: 0x0000000000000004 Type: Player Low: 4) - SMSG_MOVE_UNSET_CAN_FLY Set 
2023-03-30_12:57:14 INFO  [anticheat] AnticheatMgr:: Speed-Hack (Speed Movement at 20 above allowed Server Set rate 7.) detected player Yarillee (GUID Full: 0x0000000000000004 Type: Player Low: 4) - Latency: 1 ms - Cheat Flagged at: .go xyz -8849.628906 729.678650 99.732063 0 4.953861
2023-03-30_12:57:14 INFO  [anticheat] AnticheatMgr:: Teleport-Hack detected player Yarillee (GUID Full: 0x0000000000000004 Type: Player Low: 4) - Latency: 1 ms - Cheat Flagged at: .go xyz -8598.000000 527.000000 109.000000 0 4.953861
2023-03-30_12:57:14 INFO  [anticheat] ANTICHEAT COUNTER MEASURE:: Yarillee TELEPORT HACK REVERTED PLAYER BACK TO .go xyz -8849.628906 729.678650 99.732063 0 4.953861
```

FOR PEOPLE UPDATING FROM CURRENT. RUN THIS SQL ON THE CHARACTERS DB:
```
-- run only if updating from previous table versions
ALTER TABLE `daily_players_reports`
	ADD COLUMN `counter_measures_reports` BIGINT UNSIGNED NOT NULL DEFAULT 0 AFTER `op_ack_hack_reports`;
	
ALTER TABLE `players_reports_status`
	ADD COLUMN `counter_measures_reports` BIGINT UNSIGNED NOT NULL DEFAULT 0 AFTER `op_ack_hack_reports`;
```

**CONF:**
```

###################################################################################################
#   ANTICHEAT COUNTER MEASURE
#
#     Anticheat Counter Measures
#
#     Anticheat.CM.TIMEMANIPULATION
#
#       Description: Counters Time Manipulation hacks. These hacks can cause server to crash.
#
#       Default:    1 - (Enabled)
#                   0 - (Disabled)
#

Anticheat.CM.TIMEMANIPULATION = 1

#
#     Anticheat.CM.Teleport
#
#       Description: Sends player back to last gps position if player cheats and teleports.
#
#       Default:    0 - (Disabled)
#                   1 - (Enabled)
#

Anticheat.CM.Teleport = 0

#
#     Anticheat.CM.FLYHACK
#
#       Description: Sets player back to the ground if fly hack is detected
#
#       Default:    0 - (Disabled)
#                   1 - (Enabled)
#

Anticheat.CM.FLYHACK = 0
#
#     Anticheat.CM.JUMPHACK
#     Anticheat.CM.ADVJUMPHACK
#       Description: Sets player back to the ground if jump hack is detected
#
#       Default:    0 - (Disabled)
#                   1 - (Enabled)
#

Anticheat.CM.JUMPHACK = 0
Anticheat.CM.ADVJUMPHACK = 0

#
#     Anticheat.CM.IGNOREZ
#
#       Description: Sets player back to the ground if ignore-z hack is detected
#
#       Default:    0 - (Disabled)
#                   1 - (Enabled)
#

Anticheat.CM.IGNOREZ = 0

#
#     Anticheat.CM.SPEEDHACK
#
#       Description: Sets player back to allowed server rate speed when speed hack is detected
#       Default:    0 - (Disabled)
#                   1 - (Enabled)
#

Anticheat.CM.SPEEDHACK = 0

#
#     Anticheat.CM.WriteLog
#
#       Description: Enables Logging
#       Default:    0 - (Disabled)
#                   1 - (Enabled)
#

Anticheat.CM.WriteLog = 0

#
#     Anticheat.CM.ALERTSCREEN
#
#       Description: Alerts online GMs of Counter Measure being deployed via screen alert
#       Default:    0 - (Disabled)
#                   1 - (Enabled)
#

Anticheat.CM.ALERTSCREEN = 0

#
#     Anticheat.CM.ALERTCHAT
#
#       Description: Alerts online GMs of Counter Measure being deployed via chat alert
#       Default:    0 - (Disabled)
#                   1 - (Enabled)
#

Anticheat.CM.ALERTCHAT= 0
```
**Target branch(es)**: 335-passive_anticheat

**Issues addressed**: Fixes #

**Tests performed**: (Does it build, tested in-game, etc)

**Known issues and TODO list**:

- [ ] 
- [ ]

### CODE REVIEW

**Comment by acidmanifesto** [comment]
@jackpoz review please, but do not merge.

File: src/server/game/Anticheat/AnticheatMgr.cpp [hunk 1]
```diff
@@ -223,29 +226,50 @@ void AnticheatMgr::SpeedHackDetection(Player* player, MovementInfo movementInfo)
...
     {
-        if (sWorld->getBoolConfig(CONFIG_ANTICHEAT_WRITELOG_ENABLE))
+        if (sWorld->getBoolConfig(CONFIG_ANTICHEAT_CM_WRITELOG))
         {
-            TC_LOG_INFO("anticheat.module", "ANTICHEAT COUNTER MEASURE:: %s Time Diff Corrected(Map: %u) (possible Zero Time Manipulation)", player->GetName().c_str(), player->GetMapId());
+            TC_LOG_INFO("anticheat", "ANTICHEAT COUNTER MEASURE:: %s Time Diff Corrected(Map: %u) (possible Zero Time Manipulation)", player->GetName().c_str(), player->GetMapId());
+        }
+        if (sWorld->getBoolConfig(CONFIG_ANTICHEAT_CM_ALERTSCREEN))
+        {   // display warning at the center of the screen, hacky way?
+            std::string str = "|cFFFFFC00[Playername:|cFF00FFFF[|cFF60FF00" + std::string(player->GetName().c_str()) + "|cFF00FFFF] TIME MANIPULATION COUNTER MEASURE ALERT";
```
Comment (jackpoz) [inline src/server/game/Anticheat/AnticheatMgr.cpp]: isn't player->GetName() already a std::string ?

File: src/server/game/Anticheat/AnticheatMgr.cpp [hunk 2]
```diff
@@ -315,6 +369,35 @@ void AnticheatMgr::FlyHackDetection(Player* player, MovementInfo movementInfo)
...
     }
+    if (sWorld->getBoolConfig(CONFIG_ANTICHEAT_CM_FLYHACK))
+    {   // display warning at the center of the screen, hacky way?
+        std::string str = "|cFFFFFC00[Playername:|cFF00FFFF[|cFF60FF00" + std::string(player->GetName().c_str()) + "|cFF00FFFF] FLY HACK COUNTER MEASURE ALERT";
+        WorldPacket data(SMSG_NOTIFICATION, (str.size() + 1));
+        data << str;
+        sWorld->SendGlobalGMMessage(&data);
+
+        // Drop them with a op code set if they use a exploit or hack app
+        Player* cheatertarget = player->GetSession()->GetPlayer();
```
Comment (jackpoz) [inline src/server/game/Anticheat/AnticheatMgr.cpp]: isn't player->GetSession()->GetPlayer() the same as player ?

File: src/server/game/Movement/MotionMaster.cpp [hunk 3]
```diff
@@ -973,6 +973,47 @@ void MotionMaster::MoveFall(uint32 id/* = 0*/)
     Add(movement);
 }
 
+// Created a new method per advise from jackpoz instead of modifying the above
```
Comment (jackpoz) [inline src/server/game/Movement/MotionMaster.cpp]: Please remove my name 😄

**Review by jackpoz — CHANGES_REQUESTED** [review]

File: src/server/game/Anticheat/AnticheatMgr.cpp (hunk 1 above)
Comment (acidmanifesto) [reply src/server/game/Anticheat/AnticheatMgr.cpp]: updated

File: src/server/game/Anticheat/AnticheatMgr.cpp (hunk 2 above)
Comment (acidmanifesto) [reply src/server/game/Anticheat/AnticheatMgr.cpp]: correct it is.
Corected

File: src/server/game/World/World.cpp [hunk 4]
```diff
@@ -1507,6 +1507,17 @@ void World::LoadConfigSettings(bool reload)
     m_bool_configs[CONFIG_ANTICHEAT_BG_START_HACK_ENABLE] = sConfigMgr->GetBoolDefault("Anticheat.DetectBGStartHack", false);
     m_bool_configs[CONFIG_ANTICHEAT_BG_START_COUNTERHACK_ENABLE] = sConfigMgr->GetBoolDefault("Anticheat.BG.StartAreaTeleport", false);
     m_bool_configs[CONFIG_ANTICHEAT_OP_ACK_HACK_ENABLE] = sConfigMgr->GetBoolDefault("Anticheat.OpAckOrderHack", false);
+    m_bool_configs[CONFIG_ANTICHEAT_CM_TIMEMANIPULATION] = sConfigMgr->GetBoolDefault("Anticheat.CM.TIMEMANIPULATION", false);
```
Comment (jackpoz) [inline src/server/game/World/World.cpp]: usually the default values here should match the default values in worldserver.conf.dist , but I see that Anticheat.CM.TIMEMANIPULATION is enabled by default. Should it be enabled or disabled ?

File: src/server/game/Anticheat/AnticheatMgr.cpp [hunk 5]
```diff
@@ -1372,7 +1579,7 @@ void AnticheatMgr::BuildReport(Player* player, uint8 reportType)
                 TC_LOG_INFO("anticheat", "AnticheatMgr:: Reports reached assigned threshhold and counteracted by kicking player %s (%s)", player->GetName().c_str(), player->GetGUID().ToString().c_str());
             }
             // display warning at the center of the screen, hacky way?
-            std::string str = "|cFFFFFC00[Playername:|cFF00FFFF[|cFF60FF00" + std::string(player->GetName().c_str()) + "|cFF00FFFF] Auto Kicked for Reaching Cheat Threshhold!";
+            std::string str = "|cFFFFFC00[Playername:|cFF00FFFF[|cFF60FF00" + std::string(player->GetName()) + "|cFF00FFFF] Auto Kicked for Reaching Cheat Threshhold!";
```
Comment (jackpoz) [inline src/server/game/Anticheat/AnticheatMgr.cpp]:
```suggestion
            std::string str = "|cFFFFFC00[Playername:|cFF00FFFF[|cFF60FF00" + player->GetName() + "|cFF00FFFF] Auto Kicked for Reaching Cheat Threshhold!";
```

Comment (jackpoz) [reply src/server/game/Anticheat/AnticheatMgr.cpp]: Pretty much wherever you used `std::string(player->GetName())` or `std::string(player->GetName().c_str())` you should just use `player->GetName()`.

player->GetName() returns a string already so there's no need to call std::string( ) . Did this code get accepted flawlessly to another certain project ?

**Review by jackpoz — CHANGES_REQUESTED** [review]

File: src/server/game/World/World.cpp (hunk 4 above)
Comment (acidmanifesto) [reply src/server/game/World/World.cpp]: Hmm... well i was not certain if it was recommended to leave it iff by default in the core. However i do feel strongly recommend it is on by default. Should i change that in the conf or should i change up the core to have it as on?

File: src/server/game/Anticheat/AnticheatMgr.cpp (hunk 5 above)
Comment (acidmanifesto) [reply src/server/game/Anticheat/AnticheatMgr.cpp]: The workflow is set up differently, so many things thats usually accepted there shows up as errors here. But i have no issues making the changes in a bit.

File: src/server/game/World/World.cpp (hunk 4 above)
Comment (jackpoz) [reply src/server/game/World/World.cpp]: that's up to you, you can think about it more, I just want both the same setting in .conf and .cpp

Comment (acidmanifesto) [reply src/server/game/World/World.cpp]: i went ahead and adjusted it to true

**Review by jackpoz — APPROVED** [review]

**Pull request merged**